### ISSUE TITLE
`indexing.IndexQueue.process` fails to clear the queue in case of errors

### ISSUE THREAD

**Issue by d-maurer** [zopefoundation/Products.CMFCore#119]
Version 2.5.4

If an exception is raised during `process`, the queue remains unchanged.

This is dangerous: the queue is maintained globally (more precisely "thread locally"); thus, followup requests will see the queue content. The (typically used) queue processor (`PortalCatalogProcessor`) effectively looks the catalog up via utility lookup. If a followup requests targets a different portal, then the wrong catalog is updated. If the error persists (which is very likely), then the queue will never again get empty.

The problem is demonstrated by the following console session.  It is particularly nasty that the `commit` near the end does not indicate any problem but leaves the queue in a broken state.
```pycon
>>> from transaction import commit
>>> from zope.component import getSiteManager, provideAdapter
>>> 
>>> from zope.component.hooks import setSite
>>> from zope.interface import implementer
>>> from OFS.Folder import Folder
>>> 
>>> from ZODB.DemoStorage import DemoStorage
>>> from ZODB.DB import DB
>>> from App.ZApplication import ZApplicationWrapper
>>> from OFS.Application import Application
>>> 
>>> from Products.CMFCore.interfaces import IContentish
>>> from Products.CMFCore.PortalObject import PortalObjectBase
>>> from Products.CMFCore.CatalogTool import (CatalogTool, ICatalogTool, 
...      getQueue, processQueue, IIndexableObject)
>>> from Products.CMFCore.indexing import PortalCatalogProcessor, IIndexQueueProcessor
>>> from Products.CMFCore.PortalFolder import PortalFolder
>>> from Products.CMFCore.CMFCatalogAware import CatalogAware
>>> from Products.CMFCore.utils import getToolByName
>>> 
>>> @implementer(IContentish)
... class PortalFolder(PortalFolder, CatalogAware): pass
... 
>>> def setup_portal(pid):
...   """set up a portal inside a folder; return portal."""
...   f = Folder(pid + "_folder")
...   app._setObject(f.getId(), f)
...   f = app[f.getId()]
...   po = PortalObjectBase(pid)
...   f._setObject(po.getId(), po)
...   po = f[po.getId()]
...   setSite(po)
...   catalog = CatalogTool()
...   po._setObject(catalog.getId(), catalog)
...   sm = getSiteManager()
...   sm.registerUtility(po.portal_catalog, ICatalogTool)
...   sm.registerUtility(PortalCatalogProcessor(), IIndexQueueProcessor)
...   f = PortalFolder("test folder")
...   po._setObject(f.getId(), f)
...   f = po[f.getId()]
...   f.indexObject()
...   return po
... 
>>> db = DB(DemoStorage())
>>> app_wrapper = ZApplicationWrapper(db, "Application", Application)
>>> app=app_wrapper()
>>> commit()
>>> 
>>> p = setup_portal("p")
>>> f = p["test folder"]
>>> commit()
>>> 
>>> # ensure `catalog_object` fails
>>> def fail(*args): raise NotImplementedError
... 
>>> sm = getSiteManager()
>>> sm.registerAdapter(fail, (None, None), IIndexableObject)
>>> 
>>> f.indexObject()
>>> vars(getQueue())
{'queue': [(1, <PortalFolder at /p_folder/p/test folder>, None, None)], 'tmhook': <bound method QueueTM.register of <Products.CMFCore.indexing.QueueTM object at 0x7f99fbad1be8>>}
>>> commit()
>>> vars(getQueue())
{'queue': [(1, <PortalFolder at /p_folder/p/test folder>, [], None)], 'tmhook': <bound method QueueTM.register of <Products.CMFCore.indexing.QueueTM object at 0x7f99fbad1be8>>}
>>> processQueue()
Traceback (most recent call last):
 ...
NotImplementedError
>>> vars(getQueue())
{'queue': [(1, <PortalFolder at /p_folder/p/test folder>, [], None)], 'tmhook': <bound method QueueTM.register of <Products.CMFCore.indexing.QueueTM object at 0x7f99fbad1be8>>}
```

**Comment by d-maurer** [comment]
>  It is particularly nasty that the `commit` near the end does not indicate any problem but leaves the queue in a broken state.

That `commit` does not indicate the error is caused by a `transaction` bug fixed in `transaction==3.0.1`.

**Comment by d-maurer** [comment]
Use of `transaction>=3.0.1` effectively resolves the issue. It ensures that an exception from the `IndexQueue` processing is propagated out of `commit` and the (usually) resulting `abort` clears the queue.
